packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java

```diff
@@ -202,6 +187,45 @@ public ResponseEntity<Dataset> getDataset(@PathVariable("id") final UUID id) {
 		}
 	}
 
+	/**
+	 * Extracts columns from the dataset if they are not already set and saves the dataset.
+	 * @param dataset dataset to extract columns from
+	 * @return the dataset with columns extracted and saved
+	 * @throws IOException if there is an issue saving the dataset after extracting columns
+	 */
+	private Optional<Dataset> extractColumnsAsNeededAndSave(final Optional<Dataset> dataset) throws IOException {
+		if(dataset.isEmpty()){
+			//that's weird. let someone else handle this.
+			return dataset;
+		}
+
+		if(dataset.get().getColumns() != null && !dataset.get().getColumns().isEmpty()){
+			//columns are set. No need to extract
+			return dataset;
+		}
+		if(dataset.get().getFileNames() == null || dataset.get().getFileNames().isEmpty()){
+			//no file names to extract columns from
+			return dataset;
+		}
```

Comment:
Let's change that to make sure that we update the missing columns fields.
Because the CSV might be updated and therefore need to re-run this for the new or remove columns.

Reply:
These are quick failure cases - in this case here there are no files attached to the dataset (so no CSV to read headers from)

In the case above it we only continue when the column fields are empty (beaker blows away the column metadata when it does a transform now). We don't want to do this column check on every get because it is EXPENSIVE - we need to fetch the CSV file to do it